tests/unit/test_url.py

```diff
@@ -0,0 +1,22 @@
+from pip._internal.index import _clean_link
+
+_path = "/path with space/and\"[more]:"
+_clean_path = "/path%20with%20space/and%22%5Bmore%5D%3A"
+
+
+def test_hostname_clean_link():
+    url = "http://localhost.localdomain:8181" + _path
+    assert(_clean_link(url) == "http://localhost.localdomain:8181" +
+           _clean_path)
+
+
+def test_ipv4_clean_link():
```

Comment:
Rather than making a separate test for each test case, it would be better to use `@pytest.mark.parametrize()`. Then each test case can simply be an additional argument to `parametrize()`.

Reply:
Done. I am still using two extra variables, `_path` and `_clean_path`. Is that appropriate? Or should I fold that into the arguments for the decorator?